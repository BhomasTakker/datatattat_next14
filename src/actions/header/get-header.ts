"use server";

import { PATHS } from "@/lib/routing/paths";
import { cloneDeep } from "@/utils/object";
import { HeaderType } from "@/types/header";
import { getHeader } from "@/lib/mongo/actions/header";

export const trimHeader = (route: string) => {
	const header = route.split("/");
	return header.splice(0, header.length - 1).join("/");
};

export const isProtectedRoute = (route: string) => {
	//This isn't great?
	//missed a thing and got into an infinite loop
	//We CANNOT do this on Vercel!!
	// Could instead count and drop after x iterations to be safe
	return !route || route === PATHS.home() || route === "/users";
};

export async function getMainHeader(): Promise<HeaderType> {
	const header = await getHeader(PATHS.home());

	return cloneDeep(header);
}

async function populateSubHeaders(route: string): Promise<HeaderType[]> {
	if (isProtectedRoute(route)) {
		return [];
	}

	const result = await getHeader(route);
	const header = [result];

	const parentHeader = trimHeader(route);

	const sub = parentHeader ? await populateSubHeaders(parentHeader) : [];

	return [...header, ...sub] as HeaderType[];
}

export async function getSubHeaders(route: string): Promise<HeaderType[]> {
	const subHeaders = await populateSubHeaders(route);
	return cloneDeep(subHeaders);
}
