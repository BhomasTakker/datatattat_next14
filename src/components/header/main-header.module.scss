@use "../../scss/abstracts/units/" as *;
@use "../../scss/abstracts/ui/" as *;

.root {
	.divider {
		border-top: $divider;
	}

	.container {
		padding: $rem_0 $page_marginSm;
		@media screen and (min-width: $screenWidthMd) {
			padding: $rem_0 $page_marginMd;
		}

		@media screen and (min-width: $screenWidthLg) {
			padding: $rem_0 $page_marginLg;
		}

		@media screen and (min-width: $screenWidthXl) {
			padding: $rem_0 $page_marginXl;
		}

		@media screen and (min-width: $screenWidthXxl) {
			padding: $rem_0 $page_marginXxl;
		}
	}
}
