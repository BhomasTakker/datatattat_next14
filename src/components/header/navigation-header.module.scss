@use "../../scss/abstracts/colours/" as colours;

.root {
	display: flex;
	justify-content: space-between;
	align-items: center;
	padding: 1rem 0;

	border-bottom: 1px solid colours.$shadowLight2;

	.logo {
		color: colours.$logoDark;
	}

	//
	.simpleButton {
		// background-color: colours.$white;
		color: colours.$logoDark;
		padding: 0.5rem 1rem;
		border-radius: 5px;
		display: flex;
		justify-content: space-between;
		align-items: center;
		cursor: pointer;
		p {
			color: colours.$textColourDark;
		}
	}
}
