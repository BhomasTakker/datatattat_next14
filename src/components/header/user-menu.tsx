"use client";

import { PATHS } from "@/lib/routing/paths";
import Link from "next/link";
import { BiSolidUser } from "react-icons/bi";
import styles from "./user-menu.module.scss";
import { useRef } from "react";
import { useSession } from "next-auth/react";
import { usePathname } from "next/navigation";

export const UserMenu = () => {
	const { data } = useSession();
	const pathname = usePathname();
	const modalRef = useRef<HTMLDialogElement>(null);

	const openModal = () => {
		if (modalRef.current) {
			modalRef.current.showModal();
		}
	};
	const closeModal = () => {
		if (modalRef.current) {
			modalRef.current.close();
		}
	};

	return (
		<div>
			<button className={styles.menuButton} onClick={openModal}>
				<BiSolidUser size={"2rem"} />
			</button>
			<dialog ref={modalRef} className={styles.menu} onBlur={closeModal}>
				<ul>
					{/* // callback to current page */}
					<li>
						<Link href={PATHS.signOut(pathname)}>Sign Out</Link>
					</li>
					<li>
						{/* add page query? */}
						<Link href={PATHS.edit()}>Edit</Link>
					</li>
					{data?.user?.name ? (
						<li>
							<Link href={PATHS.user(data?.user?.name || "")}>User Page</Link>
						</li>
					) : null}
				</ul>
			</dialog>
		</div>
	);
};
