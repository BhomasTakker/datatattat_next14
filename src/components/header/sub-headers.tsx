import React, { ReactElement } from "react";
import { NavigationMenu } from "./navigation/navigation-menu";
import { HeaderType } from "@/types/header";

const renderSubMenu = (item: HeaderType, index: number): ReactElement => {
	const { nav = [] } = item || {};
	return <NavigationMenu items={nav} key={index} />;
};

const renderList = (list: HeaderType[]): ReactElement => {
	return <div>{list.map(renderSubMenu)}</div>;
};

const trimList = (list: HeaderType[]) => {
	const { length } = list;
	// Should use a config value?
	return length <= 2 ? list : list.slice(length - 2);
};

export const SubHeaders = ({
	headersArray,
}: {
	headersArray: HeaderType[];
}) => {
	if (!headersArray || headersArray.length === 0) {
		return null;
	}

	// Take in list and reverse
	const reversedList = headersArray.reverse();

	const trimmedList = trimList(reversedList);

	return renderList(trimmedList);
};
