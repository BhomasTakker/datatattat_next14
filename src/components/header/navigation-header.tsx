import { AiFillHome } from "react-icons/ai";
import styles from "./navigation-header.module.scss";
import Link from "next/link";
import { PATHS } from "@/lib/routing/paths";
import { getServerSession } from "next-auth";
import { options } from "@/app/api/auth/[...nextauth]/options";
import { UserMenu } from "./user-menu/user-menu";
import { BiSolidUser } from "react-icons/bi";
import { getUserById } from "@/lib/mongo/actions/user";
import { Session } from "@/types/auth/session";

const Logo = () => {
	return (
		<Link aria-label="Home" href="/">
			<div className={styles.homeContainer}>
				<AiFillHome className={styles.logo} />
			</div>
		</Link>
	);
};

const Datatattat = () => {
	return (
		<div className={styles.logoContainer}>
			<Link href={"/"} key={"Home"} aria-hidden>
				{/* Pull from state - We should have a state config */}
				<h2 className={styles.datatattat}>DATATATTAT</h2>
			</Link>
		</div>
	);
};

const SignInButton = () => {
	return (
		// callback to current page
		<Link className={styles.simpleButton} href={PATHS.signIn("/")}>
			<BiSolidUser className={styles.logo} />
			<p>Login</p>
		</Link>
	);
};

export const NavigationHeader = async () => {
	const session = (await getServerSession(options)) as Session;

	let menuComponent = <div />;

	// Warning
	// Here may look like a bug
	// If we have an active session in local storage i.e. for Development
	// If we then are running in 'Production' we will not have a valid session
	// So the user button to sign in or modify will not be shown
	if (!session) {
		menuComponent = <SignInButton />;
	} else {
		// We can have in local storage the wrong user - from Prod or Staging etc
		const { user } = session;
		const { user_id } = user;
		const userData = await getUserById(user_id);

		if (userData) {
			const { username, avatar } = userData;
			menuComponent = <UserMenu username={username} avatar={avatar} />;
		}
	}

	return (
		<nav className={styles.root}>
			<Logo />
			<Datatattat />
			{menuComponent}
		</nav>
	);
};
