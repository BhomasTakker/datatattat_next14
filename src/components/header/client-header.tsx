import { getMainHeader, getSubHeaders } from "@/actions/header/get-header";
import { NavigationMenu } from "./navigation/navigation-menu";
import { SubHeaders } from "./sub-headers";
import { DynamicHeader } from "./main-header";

type Props = {
	edit?: boolean;
} & DynamicHeader;

// Next 14 we will cache the header
// Next 15 will not
// We need to cache and bust by route
export const ClientHeader = async ({ route, edit = false }: Props) => {
	const header = await getMainHeader();
	// createRouteFromParams return `/${joined}`
	const joined = (route || []).join("/");
	const joinedRoute = `/${joined}`;
	const subHeaders =
		joinedRoute !== "/" ? await getSubHeaders(joinedRoute) : [];

	return (
		<>
			{header ? <NavigationMenu items={header.nav} edit={edit} /> : null}
			{subHeaders && subHeaders.length > 0 ? (
				<SubHeaders headersArray={subHeaders} edit={edit} />
			) : null}
		</>
	);
};
