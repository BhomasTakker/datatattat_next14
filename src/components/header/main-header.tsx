import { ClientHeader } from "./client-header";
import { NavigationHeader } from "./navigation-header";
import styles from "./main-header.module.scss";

export interface DynamicHeader {
	route?: string[] | undefined;
}
export const MainHeader = ({ route }: DynamicHeader) => {
	return (
		<header className={styles.root}>
			<NavigationHeader />
			<ClientHeader route={route} />
		</header>
	);
};
