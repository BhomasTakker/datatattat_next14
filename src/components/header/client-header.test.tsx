import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import { ClientHeader } from "./client-header";
import { getMainHeader, getSubHeaders } from "../../actions/header/get-header";

jest.mock("./navigation/navigation-menu", () => {
	return {
		NavigationMenu: ({ items }: { items: string[] }) => {
			const list = items.map((item, index) => (
				<li key={index}>
					<p>{item}</p>
				</li>
			));
			return <ul data-testid="nav">{list}</ul>;
		},
	};
});

jest.mock("./sub-headers", () => {
	return {
		SubHeaders: ({ headersArray }: { headersArray: string[] }) => {
			const menu = headersArray.map((item, index) => (
				<li key={index}>
					<p>{item}</p>
				</li>
			));
			return <ul data-testid="sub-headers">{menu}</ul>;
		},
	};
});

// second timne we mock this guy - find a good solution
jest.mock("next/navigation", () => {
	return {
		usePathname: jest.fn().mockResolvedValue("/mock-route"),
	};
});

// we can have a mocks folder but - can we call a shortcut finction to mock the module?
jest.mock("../../actions/header/get-header", () => {
	return {
		getSubHeaders: jest
			.fn()
			.mockResolvedValue(["header1", "header2", "header3"]),
		getMainHeader: jest.fn().mockResolvedValue({
			route: "/mock-route",
			nav: ["nav1", "nav2", "nav3"],
		}),
	};
});

describe("Client Header", () => {
	it("renders navigation menu", async () => {
		render(<ClientHeader />);
		await waitFor(() => {
			const nav = screen.queryByTestId("nav");
			expect(nav).toBeInTheDocument();
		});
	});

	it("does not render navigation menu when no header returned", async () => {
		(getMainHeader as jest.Mock).mockResolvedValueOnce(undefined);
		render(<ClientHeader />);
		await waitFor(() => {
			const nav = screen.queryByTestId("nav");
			expect(nav).not.toBeInTheDocument();
		});
	});

	it("renders sub headers", async () => {
		render(<ClientHeader />);
		await waitFor(() => {
			const subHeaders = screen.queryByTestId("sub-headers");
			expect(subHeaders).toBeInTheDocument();
		});
	});

	it("does not render sub headers when empty array returned", async () => {
		(getSubHeaders as jest.Mock).mockResolvedValueOnce([]);
		render(<ClientHeader />);
		await waitFor(() => {
			const nav = screen.queryByTestId("sub-headers");
			expect(nav).not.toBeInTheDocument();
		});
	});

	it("renders navigation unchanged", async () => {
		const { container } = render(<ClientHeader />);
		await waitFor(() => {
			screen.queryByTestId("nav");
			screen.queryByTestId("sub-headers");
		});
		expect(container).toMatchSnapshot();
	});
});
