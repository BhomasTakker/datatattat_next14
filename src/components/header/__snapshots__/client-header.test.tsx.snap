// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Client Header renders navigation unchanged 1`] = `
<div>
  <ul
    data-testid="nav"
  >
    <li>
      <p>
        nav1
      </p>
    </li>
    <li>
      <p>
        nav2
      </p>
    </li>
    <li>
      <p>
        nav3
      </p>
    </li>
  </ul>
</div>
`;
